### PR TITLE
Update clang to 15.0.6

### PR DESCRIPTION
**Pull request by tpudlik**
I also removed clang version number strings ("12.0.0") where possible to make similar upgrades less painful in the future.  Unfortunately, there's a limit to what can be achieved here, because the file structure of clang releases is not actually stable (so just replacing the version number with a wildcard is not generally enough).

Tests don't build yet.

### CODE REVIEW

**Comment by silvergasp** [comment]
It looks like the CI failure isn't related to your PR, and Bazel has deprecated the bundling of `@platforms//...` into the main binary. Adding this to the top of the WORKSPACE file should fix that issue.

```python
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
http_archive(
    name = "platforms",
    urls = [
        "https://mirror.bazel.build/github.com/bazelbuild/platforms/releases/download/0.0.6/platforms-0.0.6.tar.gz",
        "https://github.com/bazelbuild/platforms/releases/download/0.0.6/platforms-0.0.6.tar.gz",
    ],
    sha256 = "5308fc1d8865406a49427ba24a9ab53087f17f5266a7aabbfc28823f3916e1ca",
)
```

I imagine if this works for you locally your likely using an older version of bazel, than what's being used on the CI machine.

**Comment by tpudlik** [comment]
I actually ran into the same issue when upgrading Pigweed to bazel 6.0.0 (point 2 in https://g-issues.pigweed.dev/issues/259298367#comment5). I didn't see this locally, I think because the deprecated usage is in the `com_google_googletest` repo [only pulled in by CI](https://github.com/bazelembedded/rules_cc_toolchain/blob/main/third_party/test_repos.bzl#L16).

Is there a simple way to run the CI tests locally, to make sure I updated everything that needs to be updated?

**Comment by silvergasp** [comment]
Yeah, there aren't really many local tests. The "tests" for this repository are to build and run tests from a few well-known repositories, to catch any compiler or runtime errors. So you can run the test locally something like;
```bash
bazel build @com_google_googletest//... @com_github_google_benchmark//... @com_google_absl//...
bazel test @com_google_googletest//... @com_github_google_benchmark//... @com_google_absl//... --test_tag_filters=-benchmark
```

**Comment by tpudlik** [comment]
We're getting there! With the latest commits, most tests are good:

```
$ bazel test --keep_going @com_google_googletest//... @com_github_google_benchmark//... @com_google_absl//... --test_tag_filters=-benchmark
[snip]
Executed 273 out of 278 tests: 270 tests pass, 5 fail to build and 3 fail locally.
```

**Comment by silvergasp** [comment]
> I actually ran into the same issue when upgrading Pigweed to bazel 6.0.0 (point 2 in https://g-issues.pigweed.dev/issues/259298367#comment5). I didn't see this locally, I think because the deprecated usage is in the `com_google_googletest` repo [only pulled in by CI](https://github.com/bazelembedded/rules_cc_toolchain/blob/main/third_party/test_repos.bzl#L16).
> 
> Is there a simple way to run the CI tests locally, to make sure I updated everything that needs to be updated?

Do you have a public URL for that issue? The one you link seems to be an internal google issue.

**Comment by silvergasp** [comment]
> We're getting there! With the latest commits, most tests are good:
> 
> ```
> $ bazel test --keep_going @com_google_googletest//... @com_github_google_benchmark//... @com_google_absl//... --test_tag_filters=-benchmark
> [snip]
> Executed 273 out of 278 tests: 270 tests pass, 5 fail to build and 3 fail locally.
> ```

Oh nice!

**Comment by tpudlik** [comment]
> Do you have a public URL for that issue? The one you link seems to be an internal google issue.

Ah, sorry. Yes, it's https://issues.pigweed.dev/issues/259298367#comment5. Updated the comment above, too.

**Comment by tpudlik** [comment]
I may have upgraded googletest too far :/. The error I get when running ` bazel build @com_github_google_benchmark//test:benchmark_gtest` is,

```
external/com_google_googletest/googletest/include/gtest/internal/gtest-port.h:270:2: error: C++ versions less than C++14 are not supported.                                                                        
#error C++ versions less than C++14 are not supported. 
```

But I see the repo [explicitly asks for C++11](https://github.com/google/benchmark/blob/main/test/BUILD#L13). Indeed, I can reproduce this error by upgrading googletest in https://github.com/google/benchmark to 1.13.

I'll see if there's a compromise version of googletest that both works with bazel 6.0.0 (no deprecated platforms) and compiles with C++11.

**Comment by silvergasp** [comment]
> I may have upgraded googletest too far :/. The error I get when running ` bazel build @com_github_google_benchmark//test:benchmark_gtest` is,
> 
> ```
> external/com_google_googletest/googletest/include/gtest/internal/gtest-port.h:270:2: error: C++ versions less than C++14 are not supported.                                                                        
> #error C++ versions less than C++14 are not supported. 
> ```
> 
> But I see the repo [explicitly asks for C++11](https://github.com/google/benchmark/blob/main/test/BUILD#L13). Indeed, I can reproduce this error by upgrading googletest in https://github.com/google/benchmark to 1.13.
> 
> I'll see if there's a compromise version of googletest that both works with bazel 6.0.0 (no deprecated platforms) and compiles with C++11.

Oh yeah, I think I ran into this problem the other day. Let me see if I can track down the version I settled on.

**Comment by silvergasp** [comment]
Oh never mind, I didn't end up solving this one. Instead I dropped google benchmark as a dependency :) Maybe just drop the dependency in the CI. I'm not sure that it's adding that much value anyway.

File: third_party/clang_llvm_x86_64_linux_gnu_ubuntu.BUILD
Comment (silvergasp) [inline third_party/clang_llvm_x86_64_linux_gnu_ubuntu.BUILD:52]: As a bit of a comprimise you might be able to add a variable at the top of this build file and then just use `.format()` to allow for easy future updates. So something like this...
```python
CLANG_VERSION = "15.0.6",

cc_toolchain_import(
  name = "llvm_unwind",
  includes = ["lib/clang/{clang_version}/include".format(clang_version=CLANG_VERSION)],
  # ....
)
```

File: third_party/clang_llvm_x86_64_linux_gnu_ubuntu.BUILD
Comment (silvergasp) [inline third_party/clang_llvm_x86_64_linux_gnu_ubuntu.BUILD:117]: Is this the new location for libc++? I'm surprised it's not a stable path...

**Review by nathaniel-brough — COMMENTED** [review]
I think that with a little cleanup this would be fine. But if you are keen on making future updates easier, I've got a few suggestions;
- Identify areas where the directory structure for clang is not stable e.g. paths that contain versions and library paths;
- Create variables for them at the top of the clang build file. then format them in place :)

But otherwise this is looking good.

File: third_party/clang_llvm_x86_64_linux_gnu_ubuntu.BUILD
Comment (tpudlik) [reply third_party/clang_llvm_x86_64_linux_gnu_ubuntu.BUILD:117]: Sadly, so it seems :/. `find -L . -name libc++abi.a` finds only this one file.

File: third_party/clang_llvm_x86_64_linux_gnu_ubuntu.BUILD
Comment (tpudlik) [reply third_party/clang_llvm_x86_64_linux_gnu_ubuntu.BUILD]: Done!

**Comment by tpudlik** [comment]
Looks like we're having some CI failures caused by https://github.com/orgs/community/discussions/45830. Might need to wait a little while for those to clear.

**Comment by silvergasp** [comment]
Oof that's a doozy

**Comment by tpudlik** [comment]
NB I tried using this new clang version to build Pigweed. Mostly it works, but unfortunately I ran into many building warnings (some treated as errors) in direct (boringssl) and indirect (zlib, via rules_proto_grpc; old absl, via rules_fuzzing) dependencies. My least favorite is the large number of warnings like,

```
external/zlib/compress.c:22:13: warning: a function definition without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]          
int ZEXPORT compress2 (dest, destLen, source, sourceLen, level) 
```

that appear in zlib 1.2 (since they suggest we won't be able to build with `-std=c++20`, and simply silencing them is not an option). We're unfortunately stuck with `rules_proto_grpc` for now (https://issues.pigweed.dev/issues/234873954).

**Comment by silvergasp** [comment]
FYI, you can usually re-trigger checks by closing and opening the PR.

**Comment by silvergasp** [comment]
> NB I tried using this new clang version to build Pigweed. Mostly it works, but unfortunately I ran into many building warnings (some treated as errors) in direct (boringssl) and indirect (zlib, via rules_proto_grpc; old absl, via rules_fuzzing) dependencies. My least favorite is the large number of warnings like,
> 
> ```
> external/zlib/compress.c:22:13: warning: a function definition without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]          
> int ZEXPORT compress2 (dest, destLen, source, sourceLen, level) 
> ```
> 
> that appear in zlib 1.2 (since they suggest we won't be able to build with `-std=c++20`, and simply silencing them is not an option). We're unfortunately stuck with `rules_proto_grpc` for now (https://issues.pigweed.dev/issues/234873954).

So just to confirm adding `-Wnodeprecated-non-prototype` won't silence the warning? Also that error seems to be related to C20 not C++20. I wonder if it's possible to build for C++20 but then use C11 or something...

**Comment by tpudlik** [comment]
I'm pretty sure they've rolled back by now. But maybe I updated the checksum between the roll-forward and the rollback? I'll update them to match whatever I get when downloading the tarballs locally.

**Comment by tpudlik** [comment]
> So just to confirm adding `-Wnodeprecated-non-prototype` won't silence the warning? Also that error seems to be related to C20 not C++20. I wonder if it's possible to build for C++20 but then use C11 or something...

:facepalm: I missed that this was C20, not C++20. So maybe it's not such a big deal.

But it's still a little inconvenient, because the build rules for zlib are in `@rules_proto_grpc//third_party:BUILD.bazel.zlib`. So IIUC we'd need to add the  `-Wnodeprecated-non-prototype` there. Since we're not tracking their master branch (we're on 3.1.1 because we use transitive compilation, removed in 4.0), we'd have to create a fork to do this (or get the maintainers to create a 3.1.2 release with a cherrypick, but that seems unlikely).

**Comment by silvergasp** [comment]
> NB I tried using this new clang version to build Pigweed. Mostly it works, but unfortunately I ran into many building warnings (some treated as errors) in direct (boringssl) and indirect (zlib, via rules_proto_grpc; old absl, via rules_fuzzing) dependencies. My least favorite is the large number of warnings like,
> 
> ```
> external/zlib/compress.c:22:13: warning: a function definition without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]          
> int ZEXPORT compress2 (dest, destLen, source, sourceLen, level) 
> ```
> 
> that appear in zlib 1.2 (since they suggest we won't be able to build with `-std=c++20`, and simply silencing them is not an option). We're unfortunately stuck with `rules_proto_grpc` for now (https://issues.pigweed.dev/issues/234873954).



> > So just to confirm adding `-Wnodeprecated-non-prototype` won't silence the warning? Also that error seems to be related to C20 not C++20. I wonder if it's possible to build for C++20 but then use C11 or something...
> 
> 🤦 I missed that this was C20, not C++20. So maybe it's not such a big deal.
> 
> But it's still a little inconvenient, because the build rules for zlib are in `@rules_proto_grpc//third_party:BUILD.bazel.zlib`. So IIUC we'd need to add the `-Wnodeprecated-non-prototype` there. Since we're not tracking their master branch (we're on 3.1.1 because we use transitive compilation, removed in 4.0), we'd have to create a fork to do this (or get the maintainers to create a 3.1.2 release with a cherrypick, but that seems unlikely).

The other option would be to copy "@rules_proto_grpc//third_party:BUILD.bazel.zlib" into "@pigweed//third_party:BUILD.bazel.zlib" and then make the additions there. So long as the "@zlib//..." workspace is instantiated **before** "@rules_proto_grpc//..." it should override it.

**Comment by tpudlik** [comment]
Looks like we're down to three tests failing in `@com_google_absl`:

```
@com_google_absl//absl/time:time_test
@com_google_absl//absl/time/internal/cctz:time_zone_format_test
@com_google_absl//absl/time/internal/cctz:time_zone_lookup_test
```

I was able to reproduce this locally (but had hoped that it was some idiosyncrasy of my machine---apparently not!). This failure doesn't occur when compiling with the previous clang version (12.0.0), at least in the Pigweed repo.

Specifically, the failure in the `time_test` is,

```
[ RUN      ] Time.Breakdown
[test_util.cc : 26] RAW: Check LoadTimeZone(name, &tz) failed: America/New_York
```

The other two tests show similar issues with loading time zone information.

**Comment by silvergasp** [comment]
Hmm interesting, I wonder if that's something to do with the rpath... Sometimes bazel will do some wacko stuff and hide a full backtrace from you. What happens if you run that test directly without bazel involved. Something like `bazel-bin/external/absl/time/internal/cctz/time_zone_format_test`. 

My hunch here is that the abseil test is attempting to dynamically load some _system_ time library, which **might** be failing if the runtime path (rpath) isn't configured correctly.

**Comment by tpudlik** [comment]
That was a good hunch: indeed, when I run `bazel-bin/external/absl/time/internal/cctz/time_zone_format_test` directly, I get a pass!

But I don't think it's related to `rpath` specifically. Rather, setting `env = {"TZDIR": "absl/time/internal/cctz/testdata/zoneinfo"},` is to blame, [here](https://github.com/abseil/abseil-cpp/blob/master/absl/time/BUILD.bazel#L86). If I remove that line, I get the test to pass both with `bazel test` and running directly from `bazel-bin`.

The problem seems to be that `absl/time/internal/cctz/testdata/zoneinfo` is a relative path, and only correct if it's interpreted as relative to the abseil repository root.

**Comment by tpudlik** [comment]
Since this is actually a bug in the Abseil BUILD rule, I think I'll go ahead and exclude these three failing tests from CI (and file a bug against Abseil). Asked a question about this on the Bazel Slack, in case someone has an easy fix: https://bazelbuild.slack.com/archives/CA31HN1T3/p1675361567083029.

**Comment by silvergasp** [comment]
Well if that's the case I'd be happy to just exclude those tests from running in the CI, as it's clearly not a problem with the toolchain configuration. To do that you can use bazel's exclusion syntax so it'd be something like this;

`bazel test -- @com_google_absl//... -@com_google_absl//absl/time/...`

Note the minus sign before the second wildcard.

The other option would be to fix the build target in abseil. Which I think should (untested) be something like this;
`env = {"TZDIR": "$(rootpath //absl/time/internal/cctz:zoneinfo)"},`

See [Make Variables](https://bazel.build/reference/be/make-variables).

**Comment by silvergasp** [comment]
> Since this is actually a bug in the Abseil BUILD file, I think I'll go ahead and exclude these three failing tests from CI (and file a bug against Abseil). Asked a question about this on the Bazel Slack, in case someone has an easy fix: https://bazelbuild.slack.com/archives/CA31HN1T3/p1675361567083029.

I think the GH comment system is a little laggy :) I didn't see this before https://github.com/bazelembedded/rules_cc_toolchain/pull/45#issuecomment-1414169360.

**Comment by tpudlik** [comment]
Filed https://github.com/abseil/abseil-cpp/issues/1378 against Abseil. Unfortunately your fix doesn't _quite_ work because the label is a filegroup containing the individual files (so `rootpaths` returns a bunch of file paths), and `TZDIR` should be the directory that contains them.

File: .github/workflows/ci.yml
Comment (silvergasp) [inline .github/workflows/ci.yml:51]: Pretty sure this should be `bazelisk --bazelrc=ci.bazelrc build --config analyser ${{ matrix.lib }}`

Comment (tpudlik) [reply .github/workflows/ci.yml]: Argh, that's right!

File: .github/workflows/ci.yml
Comment (silvergasp) [inline .github/workflows/ci.yml:37]: I'm not sure that removing this is a great idea. I'm going to re-run the CI a couple of times to see how well it does with caching. I recently refactored this to ensure that I could build the compilation tests in parallel github actions, while sharing the repository and compilation caches. If you have a cache for each seperate entry in the github matrix it quickly reach's the upper limit of 10GB, and the cache's get evicted.

it might be doable to remove this section if the cache keys are configured differently https://github.com/bazelembedded/rules_cc_toolchain/pull/45/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR44, but I'm not sure.

Getting caching right can be the difference between a 5-10min CI run compared to a 1-2hr run.

**Review by nathaniel-brough — CHANGES_REQUESTED** [review]
This is looking great! Thanks for all the effort in getting this updated. It was definitely not a trivial effort :) I've got a few things to test when it comes to CI caching, and a few suggestions before I merge.

File: .github/workflows/ci.yml
Comment (silvergasp) [reply .github/workflows/ci.yml:37]: I was wrong. This caches totally fine, and the caches build times are about 5min or so which is great!

File: third_party/clang_llvm_x86_64_linux_gnu_ubuntu.BUILD
Comment (tpudlik) [reply third_party/clang_llvm_x86_64_linux_gnu_ubuntu.BUILD]: Removed! Also implemented TODO.

File: third_party/clang_llvm_x86_64_linux_gnu_ubuntu.BUILD
Comment (tpudlik) [reply third_party/clang_llvm_x86_64_linux_gnu_ubuntu.BUILD]: Removed.

File: README.md
Comment (tpudlik) [reply README.md]: Done.

**Comment by tpudlik** [comment]
OK, I think this is ready to be merged!

**Comment by silvergasp** [comment]
Looks great thanks!

**Comment by silvergasp** [comment]
bors +1

**Pull request merged**
